doc/poddoc/man/sstop.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SSTOP 1"
.TH SSTOP 1 "2008-07-12" "GNUspool Release 23" "GNUspool Print Manager"
.SH "NAME"
sstop \- cancel spooler
.PP
ssuspend \- suspend assignment of printers
.PP
srelease \- release assignment of printers
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBsstop\fR
[ \fI\-y\fR ]
.PP
\&\fBssuspend\fR
[ \fI\-y\fR ]
[ \fIn\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBsstop\fR is a program which should be invoked to bring the \fBGNUspool\fR
spooler system to an orderly halt prior to system shutdown. Any jobs
in progress will be saved, and the current page noted for resumption
when the spooler is restarted. Confirmation is requested (reply \f(CW\*(C`y\*(C'\fR
or \f(CW\*(C`n\*(C'\fR) unless the \f(CW\*(C`\-y\*(C'\fR option is specified.
.PP
\&\fBssuspend\fR halts the assignment of jobs to printers for a short time,
which defaults to 5 minutes or as given as a time in seconds last
argument to the command, or until released by \fBsrelease\fR. Again
confirmation is requested unless the \f(CW\*(C`\-y\*(C'\fR option is given.
.PP
Only a user with the \fIStop spooler\fR privilege may successfully invoke
any of these commands.
.PP
The reason why \fBssuspend\fR this may be useful is that in a network
scenario, where a new host is brought on line, a considerable amount
of network traffic is created, and the assignment of printers to jobs
and the broadcasting of information between hosts can flood the
network to an unacceptable degree.
.SH "OPTIONS"
.IX Header "OPTIONS"
Note that there is no means of saving the \f(CW\*(C`\-y\*(C'\fR option in a \fI.gnuspool\fR
file as this is not considered to be sensible. If you have to do this,
put the command in a shell \*(L"wrapper\*(R".
.IP "\-y" 4
.IX Item "-y"
Do not ask for confirmation.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.IP "\s-1SPRESTCONF\s0" 20
.IX Item "SPRESTCONF"
path name of alternative message file.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIspstart\fR\|(1),
\&\fIgnuspool.conf\fR\|(5),
\&\fIgnuspool.hosts\fR\|(5),
\&\fIspd\fR\|(8),
\&\fIspdinit\fR\|(8),
\&\fIspshed\fR\|(8).
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
Various diagnostics may be issued, read as required from the message
file \fIrest.help\fR.
.PP
The most important ones are that it is not running and that the user
is not permitted to invoke the command.
.SH "AUTHOR"
.IX Header "AUTHOR"
John M Collins, Xi Software Ltd.


src/sstop.c
/* sstop.c -- stop spooler tidily

   Copyright 2008 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include "config.h"
#include <stdio.h>
#include <sys/types.h>
#ifdef	HAVE_FCNTL_H
#include <fcntl.h>
#endif
#include <sys/ipc.h>
#include <sys/msg.h>
#include <errno.h>
#include <ctype.h>
#include "errnums.h"
#include "defaults.h"
#include "files.h"
#include "network.h"
#include "spq.h"
#include "spuser.h"
#include "ecodes.h"
#include "ipcstuff.h"
#include "incl_unix.h"
#include "incl_ugid.h"
#include "cfile.h"
#ifdef	SHAREDLIBS
#include "xfershm.h"
#include "q_shm.h"
#include "displayopt.h"
#endif

FILE	*Cfile;
int	Ctrl_chan;

uid_t	Daemuid,
	Effuid,
	Realuid;

#ifdef	SHAREDLIBS
struct	jshm_info	Job_seg;
struct	pshm_info	Ptr_seg;
struct	xfershm		*Xfer_shmp;
#ifndef	USING_FLOCK
int	Sem_chan;
#endif
DEF_DISPOPTS;
#endif

#define	DEFAULT_SUSTIME	300	/* 5 minutes */
#define	MAX_SUSTIME	7200	/* 2 hours */

void	nomem(void)
{
	fprintf(stderr,"Ran out of memory\n");
	exit(E_NOMEM);
}

void	confirm_op(const int promptnum, const int errnum)
{
	char	*cp, *prompt, string[80];

	prompt = gprompt(promptnum);
	fputs(prompt, stdout);
	fflush(stdout);
	fgets(string, sizeof(string), stdin);
	cp = string;
	while  (*cp  &&  !isalpha(*cp))
		cp++;
	if  (toupper(*cp) != 'Y')  {
		print_error(errnum);
		exit(0);
	}
}

/* Ye olde main routine. No arguments are expected.  */

MAINFN_TYPE	main(int argc, char **argv)
{
	int	countdown = 5;
	struct	spr_req	oreq;
	struct	spdet	*mypriv;
#if	defined(NHONSUID) || defined(DEBUG)
	int_ugid_t	chk_uid;
#endif

	versionprint(argv, "$Revision: 1.1 $", 0);

	if  ((progname = strrchr(argv[0], '/')))
		progname++;
	else
		progname = argv[0];
	init_mcfile();

	Realuid = getuid();
	Effuid = geteuid();
	INIT_DAEMUID;
	Cfile = open_cfile(MISC_UCONFIG, "rest.help");
	SCRAMBLID_CHECK
	SWAP_TO(Daemuid);
	mypriv = getspuser(Realuid);

	if  (Realuid != ROOTID  &&  Realuid != Daemuid  &&  !(mypriv->spu_flgs & PV_SSTOP))  {
		print_error($E{No sstop priv});
		exit(E_NOPRIV);
	}

	/* Open control MSGID */

	if  ((Ctrl_chan = msgget(MSGID, 0)) < 0)  {
		print_error($E{Spooler not running});
		exit(E_NOTRUN);
	}

	BLOCK_ZERO(&oreq, sizeof(oreq));
	oreq.spr_mtype = MT_SCHED;
	oreq.spr_un.o.spr_pid = getpid();

	if  (strcmp(progname, "ssuspend") == 0)  {
		int	hadconf = 0, sustime = 0;
		if  (argc > 1)  {
			if  (argc > 2)  {
				if  (argc > 3  ||  argv[1][0] != '-'  ||  toupper(argv[1][1]) != 'Y')  {
				ssu:
					print_error($E{Sstop ssuspend usage});
					exit(E_USAGE);
				}
				hadconf++;
				if  (!isdigit(argv[2][0]))
					goto  ssu;
				sustime = atoi(argv[2]);
			}
			if  (argv[1][0] == '-')  {
				if  (toupper(argv[1][1]) != 'Y')
					goto  ssu;
				hadconf++;
			}
			else  {
				if  (!isdigit(argv[1][0]))
					goto  ssu;
				sustime = atoi(argv[1]);
			}
		}
		if  (!hadconf)
			confirm_op($P{Confirm suspend scheduling}, $E{Sstop ssuspend not suspended});
		if  (sustime <= 0)  {
			char	*prompt = gprompt($P{Scheduling suspension time}), *cp;
			char	string[80];
			fputs(prompt, stdout);
			fflush(stdout);
			fgets(string, sizeof(string), stdin);
			cp = string;
			while  (*cp  &&  !isdigit(*cp))
				cp++;
			if  (*cp)
				sustime = atoi(cp);
			if  (sustime <= 0  ||  sustime > MAX_SUSTIME)
				sustime = DEFAULT_SUSTIME;
		}
		oreq.spr_un.o.spr_act = SO_SUSPSCHED;
		oreq.spr_un.o.spr_arg1 = sustime;
	}
	else  if  (strcmp(progname, "srelease") == 0)
		oreq.spr_un.o.spr_act = SO_UNSUSPEND;
	else  {
		if  (argc > 1)  {
			if  (argc != 2 || argv[1][0] != '-'  ||  (argv[1][1] != 'y'  && argv[1][1] != 'Y') || argv[1][2] != '\0')  {
				print_error($E{sstop usage});
				exit(E_USAGE);
			}
		}
		else
			confirm_op($P{Confirm stop spooler}, $E{sstop not stopped});
		oreq.spr_un.o.spr_act = SO_SSTOP;
		oreq.spr_un.o.spr_arg1 = $E{Sched normal stop};
	}

	while  (msgsnd(Ctrl_chan, (struct msgbuf *) &oreq, sizeof(struct sp_omsg), IPC_NOWAIT) < 0)  {
		if  (errno != EAGAIN)  {
			print_error($E{IPC msg q error});
			exit(E_SETUP);
		}
		print_error($E{IPC msg q full});
		if  (--countdown < 0)  {
			print_error($E{Sstop given up});
			exit(E_SETUP);
		}
		print_error($E{Sstop waiting});
		sleep(5);
	}
	return  0;
}
